druid/src/text/movement.rs

```diff
@@ -37,44 +43,98 @@ pub enum Movement {
     EndOfDocument,
 }
 
-/// Compute the result of movement on a selection .
-pub fn movement(m: Movement, s: Selection, text: &impl EditableText, modify: bool) -> Selection {
-    let offset = match m {
+/// Compute the result of movement on a selection.
+///
+/// returns a new selection representing the state after the movement.
+///
+/// If `modify` is true, only the leading edge (the 'end') of the selection
+/// should be changed; this is the case when the user moves with the shift
+/// key pressed.
+pub fn movement<T: EditableText + TextStorage>(
+    m: Movement,
+    s: Selection,
+    layout: &TextLayout<T>,
+    modify: bool,
+) -> Selection {
+    let (text, layout) = match (layout.text(), layout.layout()) {
+        (Some(text), Some(layout)) => (text, layout),
+        _ => {
+            debug_assert!(false, "movement() called before layout rebuild");
+            return s;
+        }
+    };
+
+    let (offset, h_pos) = match m {
         Movement::Left => {
             if s.is_caret() || modify {
-                text.prev_grapheme_offset(s.end).unwrap_or(0)
+                text.prev_grapheme_offset(s.end)
+                    .map(|off| (off, None))
+                    .unwrap_or((0, s.h_pos))
             } else {
-                s.min()
+                (s.min(), None)
             }
         }
         Movement::Right => {
             if s.is_caret() || modify {
-                text.next_grapheme_offset(s.end).unwrap_or(s.end)
+                text.next_grapheme_offset(s.end)
+                    .map(|off| (off, None))
+                    .unwrap_or((s.end, s.h_pos))
             } else {
-                s.max()
+                (s.max(), None)
             }
         }
 
-        Movement::PrecedingLineBreak => text.preceding_line_break(s.end),
-        Movement::NextLineBreak => text.next_line_break(s.end),
+        Movement::Up => {
+            let cur_pos = layout.hit_test_text_position(s.end);
+            if cur_pos.line == 0 {
+                (0, Some(cur_pos.point.x))
```

Comment:
I think this is the behavior of our existing code, but I think `Some(s.h_pos.unwrap_or(cur_pos.point.x))` is slightly more precise. And that might lead to logic simplification because I think it's the same in both branches.

Note that the code as you have it matches xi2, so I certainly can't complain :)

Reply:
Good call, I think I prefer preserving the existing h_pos if possible.